### PR TITLE
Nuke fixes

### LINKED ISSUE #389: minimega: multiple missing components in nuke
At the very least:

- nuke attempts to clean up /sys/fs/cgroup... which we don't use anymore
- nuke doesn't correctly unmount overlayfs mounts
- nuke doesn't appear to be cleaning up mega_tap instances (it appears to delete them but not remove them from bridges)

### PR DESCRIPTION
**Pull request by cdshann**
Fixes #389 

containerNuke should now properly unmount cgroups
containerNuke performs some sanity checking before nuking, as every host might not be running containers

cliNuke now grabs the tap reaper lock before making any changes to bridges / taps
cliNuke no longer calls updateBridgeInfo, as this would only overwrite information leftover from a crash
cliNuke no longer calls teardown(). functions needed to clear state after a crash were moved to nukeState()

### CODE REVIEW

**Comment by cdshann** [comment]
Don't merge me yet - adding a quick improvement.

**Comment by floren** [comment]
Are you ready to merge?

**Comment by cdshann** [comment]
Yeah this is ready.

**Comment by djfritz** [comment]
I'll try to test this today.

**Comment by djfritz** [comment]
Started 6 containers and 5 kvm instances, all on vlan 100, force killed minimega, and then restarted with -force and issued nuke. Got the following errors and most of nuke didn't work:

- 4 taps remained (make sure vms are killed before attempting to remove taps)
- container mountpoints are still mounted
- cgroup is unmounted (correct)
- minimega binary got deleted (incorrect)

.

    minimega$ nuke
    2016/02/22 22:00:47 ERROR container.go:1445: remove blkio.io_merged: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove blkio.io_merged_recursive: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove blkio.io_queued: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove blkio.io_queued_recursive: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove blkio.io_service_bytes: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove blkio.io_service_bytes_recursive: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove blkio.io_service_time: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove blkio.io_service_time_recursive: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove blkio.io_serviced: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove blkio.io_serviced_recursive: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove blkio.io_wait_time: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove blkio.io_wait_time_recursive: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove blkio.leaf_weight: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove blkio.leaf_weight_device: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove blkio.reset_stats: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove blkio.sectors: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove blkio.sectors_recursive: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove blkio.throttle.io_service_bytes: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove blkio.throttle.io_serviced: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove blkio.throttle.read_bps_device: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove blkio.throttle.read_iops_device: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove blkio.throttle.write_bps_device: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove blkio.throttle.write_iops_device: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove blkio.time: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove blkio.time_recursive: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove blkio.weight: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove blkio.weight_device: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove cgroup.clone_children: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove cgroup.event_control: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove cgroup.procs: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove cgroup.sane_behavior: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove cpu.cfs_period_us: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove cpu.cfs_quota_us: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove cpu.shares: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove cpu.stat: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove cpuacct.stat: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove cpuacct.usage: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove cpuacct.usage_percpu: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove cpuset.cpu_exclusive: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove cpuset.cpus: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove cpuset.effective_cpus: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove cpuset.effective_mems: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove cpuset.mem_exclusive: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove cpuset.mem_hardwall: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove cpuset.memory_migrate: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove cpuset.memory_pressure: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove cpuset.memory_pressure_enabled: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove cpuset.memory_spread_page: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove cpuset.memory_spread_slab: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove cpuset.mems: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove cpuset.sched_load_balance: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove cpuset.sched_relax_domain_level: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove devices.allow: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove devices.deny: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove devices.list: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove memory.failcnt: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove memory.force_empty: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove memory.limit_in_bytes: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove memory.max_usage_in_bytes: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove memory.move_charge_at_immigrate: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove memory.numa_stat: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove memory.oom_control: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove memory.pressure_level: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove memory.soft_limit_in_bytes: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove memory.stat: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove memory.swappiness: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove memory.usage_in_bytes: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove memory.use_hierarchy: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove net_cls.classid: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove net_prio.ifpriomap: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove net_prio.prioidx: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove notify_on_release: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove release_agent: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1445: remove tasks: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1468: overlay unmount: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1468: overlay unmount: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1468: overlay unmount: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1468: overlay unmount: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1468: overlay unmount: no such file or directory
    2016/02/22 22:00:47 ERROR container.go:1468: overlay unmount: no such file or directory
    2016/02/22 22:00:49 ERROR nuke.go:185: mega_tap5 -- ip: exit status 1: ioctl(TUNSETIFF): Device or resource busy

    2016/02/22 22:00:49 ERROR nuke.go:185: mega_tap6 -- ip: exit status 1: ioctl(TUNSETIFF): Device or resource busy

    2016/02/22 22:00:49 ERROR nuke.go:185: mega_tap7 -- ip: exit status 1: ioctl(TUNSETIFF): Device or resource busy

    2016/02/22 22:00:49 ERROR nuke.go:185: mega_tap8 -- ip: exit status 1: ioctl(TUNSETIFF): Device or resource busy

    2016/02/22 22:00:50 ERROR nuke.go:90: remove /tmp/minimega/4/fs/lib64: directory not empty

**Comment by djfritz** [comment]
@cdshann

**Comment by cdshann** [comment]
Thanks for the feedback. Looking into it.

**Comment by cdshann** [comment]
This should be good to go if anyone wants to do any additional testing. This works per @djfritz and I's testing earlier today.

File: src/minimega/container.go [hunk 1]
```diff
@@ -1425,67 +1426,81 @@ func containerMountDefaults(fsPath string) error {
...
-		log.Debug("walking file: %v", path)
-
-		switch info.Name() {
-		case "tasks":
-			d, err := ioutil.ReadFile(path)
-			pids := strings.Fields(string(d))
-			for _, pid := range pids {
-				log.Debug("found pid: %v", pid)
+	// Allow udev to sync
+	time.Sleep(time.Second * 1)
```
Comment (djfritz) [inline src/minimega/container.go]: I can't think of a better solution to this, but UGHHH

**Comment by djfritz** [comment]
LGTM

@floren can you review as well? It's big enough that it should have some more eyes on it.

File: src/minimega/container.go (hunk 1 above)
Comment (jcrussell) [inline src/minimega/container.go]: `udevadm settle`?

Comment (djfritz) [inline src/minimega/container.go]: Doesn't exist anymore with systemd. Joy!

Comment (djfritz) [inline src/minimega/container.go]: Correction. It does still exist. But not without systemd. Our setup won't
have it, and honestly I'm not ready to cross the systemd rubicon until we
absolutely have to.

@floren

**Comment by djfritz** [comment]
Well I still think LGTM. Merging.

**Pull request merged**